bpo-17535: IDLE editor line numbers

taleinat opened the pull request:
Line numbers for IDLE editor windows!

This is based on Saimadhav Heblikar's last patch posted to the related issue.

<!-- issue-number: [bpo-17535](https://bugs.python.org/issue17535) -->
https://bugs.python.org/issue17535
<!-- /issue-number -->

taleinat committed:
apply Saimadhav's patch v3 (linenumber-text-widget-v3.diff)

taleinat committed:
updated into first working version

* refactored from an extension into an integral part of IDLE
* got tests working
* various required fixes
* some code cleanup and stylistic changes

taleinat committed:
some refactoring

taleinat committed:
fix integration with undo/redo

taleinat committed:
give get_end a better name and a doc-string

taleinat committed:
fix buggy insertion of a delegator into the middle of the percolator

taleinat committed:
fix bug where BaseSideBar.show_sidebar() ignored the given side

also refactor the state of BaseSideBar and LineNumbers

taleinat committed:
add a NEWS entry

taleinat committed:
update line numbers immediately upon highlighting config changes

taleinat committed:
optimize and improve scrolling

* avoid duplicate scrolling events which degrade performance
* reorganize event forwarding code
* dragging on line numbers now selects entire lines
* work around difficult to handle drag-scrolling

taleinat committed:
add a separate configuration setting for line number colors

taleinat committed:
remove double- and triple- click conversion to single clicks

This is apparently unnecessary.

taleinat committed:
replace hack to make font changes take effect immediately

Taking advantage of the tighter integration with EditorWindow,
each window now directly updates its line numbers sidebar upon
font changes.

taleinat committed:
refactor and expand tests, also fixing test_delete

taleinat committed:
attempt to fix tests failing on Linux

There is apparently a minor inconsistency in Tk's Text widget.

terryjreedy reviewed (changes requested):
See separate comments.  Will look at code later.

taleinat committed:
remove sidebar text widget border on macOS

taleinat committed:
mouse click and drag now select entire lines

taleinat committed:
fix layout when showing both line numbers and code context

taleinat committed:
fix test_toggle_linenumbering

taleinat committed:
fix code context tests after pack->grid change

terryjreedy committed:
Merge remote-tracking branch 'origin/master' into pr_14030

terryjreedy committed:
10. Make text_frame resizeable.

terryjreedy committed:
Remove unneeded addition to config-extensions.def.

terryjreedy committed:
1. Skip warning for new theme elements.

terryjreedy committed:
Remane linenumbers.py to sidebar.py.

taleinat committed:
rename test class

terryjreedy committed:
Rename test_linenumbers to test_sidebar, fix import.

taleinat committed:
forward double- and triple-clicks events as normal click events

taleinat commented on Lib/idlelib/config.py:
```diff
@@ -345,11 +349,11 @@ def GetThemeDict(self, type, themeName):
-            if not cfgParser.has_option(themeName, element):
+            if (not cfgParser.has_option(themeName, element) and
+                # Skip warning for new elements.
+                element[:element.index('-')] not in ['context', 'sidebar']):
```
```suggestion
                not element.startswith(('context-', 'linenumber-')):
```

terryjreedy replied:
I agree, but cannot commit thru this interface.  Says "You do not have permission to push to this repository.", which is wrong for this branch.  Can you accept your own suggestion?

taleinat committed:
line numbers off by default and set separately per window

This is identical to how code context currently works, using the
same implementation method.

taleinat committed:
configurable default for whether to show line numbers in new windows

taleinat committed:
fix state of menu option when opening IDLE with line numbers shown

This also improves the separation of concerns between the editor
window code and the sidebar code.

taleinat committed:
fix and improve show_sidebar() and hide_sidebar() and their tests

taleinat committed:
add tests for click and drag selections

taleinat committed:
add a test for scrolling

taleinat committed:
add an htest for drag-scrolling on the line numbers

taleinat committed:
fix string formatting in htest

taleinat committed:
fix skipping warning for missing line number highlight config

taleinat committed:
fix skipping warning for missing line number highlight config

taleinat committed:
fix click and drag selection

taleinat committed:
Merge branch 'master' into bpo-17535/IDLE-line-numbers

# Conflicts:
#	Lib/idlelib/codecontext.py
#	Lib/idlelib/editor.py
#	Lib/idlelib/outwin.py

terryjreedy committed:
Merge branch 'master' into bpo-17535/IDLE-line-numbers

terryjreedy reviewed (changes requested):
I fixed pending merge conflicts.  Both CI and my manual tests of code context pass.  Cell00 looks fine, as does unfilled cell 0, 2, above scroller.  As for line numbers, line selection works as requested and crazy (additional) flashing is gone, at least on Windows.

Minor issue.  Line numbers are shifted up one pixel, at least on my machine.  I thought they seemed off, but verified and measured.  Open a new file with line numbers.  On lines 2 and 7, enter multiple 2s and 7s.  Use paper to trace underbars and overbars to sidebar.  Is there a difference in padding between the two text widgets?

Major issue: open settings; change font, apply changes.  Usually text changes, numbers do not; this is a bug.  Hit Ok.  Usually, number catch up, but one can instead exit with Cancel, which only cancels changes since apply.  'Bold []' changes act differently from size and face changes (and the latter can affect size).   The result is that after closing the config dialog, the number font may not match the text font.  This is especially bad when the spacing is different. I will try to investigate.

terryjreedy committed:
Move line numbers down a pixel.

taleinat committed:
properly ensure correct Y-axis padding of the sidebar widget

taleinat committed:
fix line numbers update upon font config changes

Also refactor the font and highlight config updating
mechanism for both line numbers and code context.

taleinat committed:
fix fragile tests in test_sidebar

terryjreedy committed:
Update test_sidebar docstring.

terryjreedy commented on Lib/idlelib/sidebar.py:
```diff
@@ -0,0 +1,323 @@
+                bind_mouse_event(event_name, target_event_name=event_name)
+
+            # Convert double- and triple-click events to normal click events,
+            # since event_generate() doesn't allow generating such events.
```
This does not mean that we cannot use double-clicks, just that we cannot test their effect, at least not easily.  But I think that this is the best thing to do until we do use them.

terryjreedy committed:
Touchup docs.

terryjreedy reviewed (approved):
Numbers are still lined up properly.  On my screen they appear to be a bit shorter.  I believe that this is the result of fg/bg color differences resulting in subpixel shading.  So it is not a bug (but might someday get reported as such).

Font changes work for me.  With this, and the doc changes I am about to push*, I am ready for this to be merged.  I have configured line numbers on so I will see them daily as soon as it is.  I will review the sidebar code and tests more closely when I try making another subclass.

* idle.rst , help.html, and the blurb.  My commit got squashed when I re-pulled to get the test_sidebar change.

terryjreedy committed:
fix locally

terryjreedy committed:
Merge branch 'bpo-17535/IDLE-line-numbers' of https://github.com/taleinat/cpython into pr_14030

taleinat committed:
more doc touch-ups: consistency, wording, typo

taleinat committed:
add tests for font and color updates for line numbers

Also cleaned up the parallel code context tests.

taleinat committed:
ensure font consistency in line numbers htest

taleinat committed:
add "What's New" entries

taleinat committed:
revert 3.9 "what's New" entry
